src/iceberg/avro/avro_data_util.cc

```diff
@@ -17,16 +17,383 @@
+      }
+
+      auto* builder =
+          internal::checked_cast<::arrow::FixedSizeBinaryBuilder*>(array_builder);
+      const auto& fixed = avro_datum.value<::avro::GenericFixed>();
+      if (fixed.value().size() != 16) {
+        return InvalidArgument("Expected UUID fixed length 16, got: {}",
+                               fixed.value().size());
+      }
+      const auto& value = fixed.value();
+      ICEBERG_ARROW_RETURN_NOT_OK(
+          builder->Append(reinterpret_cast<const uint8_t*>(value.data())));
+      return {};
+    }
+
+    case TypeId::kDecimal: {
+      if (avro_node->type() != ::avro::AVRO_FIXED ||
+          avro_node->logicalType().type() != ::avro::LogicalType::DECIMAL) {
+        return InvalidArgument(
+            "Expected Avro fixed with decimal logical type for decimal field, got: {}",
+            ToString(avro_node));
+      }
+
+      const auto& fixed = avro_datum.value<::avro::GenericFixed>();
+      const auto& value = fixed.value();
+      ICEBERG_ARROW_ASSIGN_OR_RETURN(
+          auto decimal, ::arrow::Decimal128::FromBigEndian(value.data(), value.size()));
+      auto* builder = internal::checked_cast<::arrow::Decimal128Builder*>(array_builder);
+      ICEBERG_ARROW_RETURN_NOT_OK(builder->Append(decimal));
+      return {};
+    }
+
+    case TypeId::kDate: {
+      if (avro_node->type() != ::avro::AVRO_INT ||
+          avro_node->logicalType().type() != ::avro::LogicalType::DATE) {
+        return InvalidArgument(
+            "Expected Avro int with DATE logical type for date field, got: {}",
+            ToString(avro_node));
+      }
+      auto* builder = internal::checked_cast<::arrow::Date32Builder*>(array_builder);
+      ICEBERG_ARROW_RETURN_NOT_OK(builder->Append(avro_datum.value<int32_t>()));
+      return {};
+    }
+
+    case TypeId::kTime: {
+      if (avro_node->type() != ::avro::AVRO_LONG ||
+          avro_node->logicalType().type() != ::avro::LogicalType::TIME_MICROS) {
+        return InvalidArgument(
+            "Expected Avro long with TIME_MICROS for time field, got: {}",
+            ToString(avro_node));
+      }
+      auto* builder = internal::checked_cast<::arrow::Time64Builder*>(array_builder);
+      ICEBERG_ARROW_RETURN_NOT_OK(builder->Append(avro_datum.value<int64_t>()));
+      return {};
+    }
+
+    case TypeId::kTimestamp:
+    case TypeId::kTimestampTz: {
+      if (avro_node->type() != ::avro::AVRO_LONG ||
+          avro_node->logicalType().type() != ::avro::LogicalType::TIMESTAMP_MICROS) {
+        return InvalidArgument(
+            "Expected Avro long with TIMESTAMP_MICROS for timestamp field, got: {}",
+            ToString(avro_node));
+      }
+      auto* builder = internal::checked_cast<::arrow::TimestampBuilder*>(array_builder);
+      ICEBERG_ARROW_RETURN_NOT_OK(builder->Append(avro_datum.value<int64_t>()));
+      return {};
+    }
+
+    default:
+      return InvalidArgument("Unsupported primitive type: {}",
```

Comment:
can we add this is unsupport just in avro, like `"Unsupported primitive type for avro:` ?